File: metadata/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/domain/ValidationResult.java
package io.qpointz.mill.metadata.domain;

import java.util.List;

public record ValidationResult(boolean valid, List<String> errors) {

    public static ValidationResult ok() {
        return new ValidationResult(true, List.of());
    }

    public static ValidationResult fail(String error) {
        return new ValidationResult(false, List.of(error));
    }

    public static ValidationResult fail(List<String> errors) {
        return new ValidationResult(false, errors);
    }

    public static ValidationResult merge(List<ValidationResult> results) {
        var errors = results.stream()
                .filter(r -> !r.valid())
                .flatMap(r -> r.errors().stream())
                .toList();
        return errors.isEmpty() ? ok() : fail(errors);
    }
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/domain/ValidationResult.kt
package io.qpointz.mill.metadata.domain

/** Outcome of facet or entity validation with optional error messages. */
data class ValidationResult(val valid: Boolean, val errors: List<String>) {
    companion object {
        @JvmStatic fun ok(): ValidationResult = ValidationResult(true, emptyList())
        @JvmStatic fun fail(error: String): ValidationResult = ValidationResult(false, listOf(error))
        @JvmStatic fun fail(errors: List<String>): ValidationResult = ValidationResult(false, errors)
        @JvmStatic fun merge(results: List<ValidationResult>): ValidationResult {
            val errors = results.filter { !it.valid }.flatMap { it.errors }
            return if (errors.isEmpty()) ok() else fail(errors)
        }
    }
}
